### PR TITLE
[inductor] Fuse sibling reductions with common reads

### PR DESCRIPTION
**Pull request by ahmedtaha100**
## Summary

Domino cross entropy reduces both corrected and base logits. Both reductions read the same large base tensor, but Inductor currently schedules them in separate kernels.

This PR lets Inductor fuse compatible sibling CUDA reductions when shared direct reads account for at least half of the larger reduction's input traffic. Indirect reads and disjoint slices do not qualify. The optimization has its own config switch.

Part of pytorch/pytorch#190718. Split from pytorch/pytorch#193121 at the maintainer's request.

## Results

On the exact reported BF16 shape, adding this change to the cat-recomputation patch reduced runtime from 3.686 ms to 2.642 ms and reduced the vocabulary kernels from three to two. Temporary memory remained at 0.1 MiB.

The focused CUDA test produces one kernel when common-read fusion is enabled. It produces two when disabled and two for repeated disjoint slices. All outputs match eager execution.

## Test plan

```text
python -m pip install -e . -v --no-build-isolation
python test/inductor/test_inductor_scheduler.py -k test_reduction_common_read_scoring
spin quicklint -- --skip FLAKE8 torch/_inductor/config.py torch/_inductor/scheduler.py test/inductor/test_inductor_scheduler.py
```

The focused CUDA test command passed. The scoped quicklint command also reports no issues.

## AI assistance disclosure

I wrote and reviewed the final patch and commit message. Codex assisted with tracing the scheduler behavior, suggesting parts of the implementation, running local checks, and revising this PR description. I reviewed the final diff, test results, and description, understand the changes, and take responsibility for this contribution. Thank you.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @jansel @eellison

### CODE REVIEW

**Comment by ahmedtaha100** [comment]
The Pytorch bot keeps CC'ing everyone

**Comment by eellison** [comment]
I think this is fixing the wrong abstraction. This PR adds a new 50% common-buffer heuristic, config, and extra fusion pass, but this case is missed because `cat`'s `Identity` wrapper prevents two algebraically identical reads from comparing equal. The relevant scheduler dependencies are:

```text
cat reduction:
  base[d1 + 1024*FloorDiv(d0, 16)]
  base[d1 + 64*Identity(ModularIndexing(d0, 1, 16) - 1)
          + 1024*FloorDiv(d0, 16) + 64]

base reduction:
  base[64*d0 + d1]
```

The reason we have `Identity` is to prevent int32 -> int64 expansion. It's a bit tricky, because, if we match up Memory Deps we also expect at codegen for these accesses to CSE. 

So we would need a way to align memory deps, codegen, and preserve the int64 handling.

I think this is probably not a good pr for someone relatively new to code base, sorry.

**Comment by ahmedtaha100** [comment]
@eellison Thank you for the feedback. That makes sense. Before I close this PR, could we keep Identity in the emitted index but use a shared Identity-free, range-simplified key for both MemoryDep matching and load CSE? That would let us remove the heuristic, config, and extra pass without changing the integer-width behavior. Any CSE reuse would still need to respect mask coverage. Is that the direction you had in mind? If not, I’m happy to close it.

**Comment by eellison** [comment]
can you address the other pr which i commented on first ?
